[dnf5] rpm::SolvSack loading of available repository

jrohel opened the pull request.

jrohel committed:
[dnf5] Sack load repo

jrohel committed:
[dnf5] Add unit tests for loading repository into rpm::Sack

jrohel committed:
[dnf5] unit tests: Add testing repository dnf-ci-fedora

Taken from dnf CI stack.

jrohel committed:
[dnf5] unit test for package set: use unique_ptr

jrohel committed:
[dnf5] unit test for package set: compilable with rpm::Sack changes

jrohel committed:
[dnf5] remove hard-coded nsolvables, use real repodata in tests for package set

j-mracek commented on include/libdnf/rpm/sack.hpp:
```diff
@@ -61,11 +68,53 @@ inline ReldepId::ReldepId(int id) : id(id) {}
 class ReldepList;
+class Repo;
 
 class Sack {
```
What about to name it `SolvSack`?

jrohel replied:
Yes, we agreed on the renaming to `SolvSack` some time ago. Question was who will rename it. So I'll do it in this PR.

j-mracek commented on test/libdnf/rpm/repo/test_repo.cpp:
```diff
@@ -0,0 +1,111 @@
+        log_router.error(ex.what());
+    }
+
+    /*    auto repo2 = repo_sack.new_repo("repo2");
```
Please could you explain why the code is marked as a comment.

jrohel replied:
It's code from the other test. There is no reason for it to stay here. I will remove it.

j-mracek commented on include/libdnf/rpm/sack.hpp:
```diff
@@ -61,11 +68,53 @@ inline ReldepId::ReldepId(int id) : id(id) {}
+    // @DNF_SACK_LOAD_FLAG_USE_OTHER:               Use other metadata
+    //
+    // Flags to use when loading from the sack.
+    typedef enum {
```
I believe that it could be replaced by enum class. So far it is used by repo_impl->libsolv_repo_ext.load_flags. The question is how data are passed to librepo or to libsolv?

j-mracek replied:
Please what relation between DataType and DnfSackLoadFlags?

jrohel replied:
An enum class will be better for clean C++ interface.  But it is public API. How we export it to plain C?
Anyway. The presented solution is also problematic to export. I've ideas. We must discuss it.

jrohel replied:
@j-mracek I created enum class. And done some renaming. I hope it is more clear now.

jrohel committed:
[dnf5] Rename rpm::Sack to rpm::SolvSack

jrohel committed:
[dnf5] Rename files sack* to solv_sack*

j-mracek commented on include/libdnf/rpm/package.hpp:
```diff
@@ -23,7 +23,6 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 
 #include "checksum.hpp"
 #include "reldep_list.hpp"
-#include "sack.hpp"
```
Why did you remove the import? The import was required for SolvSack and PackageId.

jrohel replied:
There is indirect include via "reldep_list.hpp". But in this case I agree with you. I will add the include back.

j-mracek commented on include/libdnf/rpm/repo.hpp:
```diff
@@ -36,6 +36,8 @@ class Base;
 
 namespace libdnf::rpm {
 
+class SolvSack;
```
In some cases we added commend that this is a forward declaration. Shall we used everywhere?

jrohel replied:
I don't know. The comment has none information value for me.
```
// forward declarations
class SolvSack;
```

For me is the same as:
```
// assign values
a = 5;
```
Maybe there was an attempt to mark one place in the code for all forward declarations. So, if you want, I can add comment "forward declarations"

j-mracek commented on include/libdnf/rpm/solv_sack.hpp:
```diff
@@ -0,0 +1,140 @@
+    std::unique_ptr<Impl> pImpl;
+};
+
+inline constexpr SolvSack::LoadRepoFlags operator|(SolvSack::LoadRepoFlags lhs, SolvSack::LoadRepoFlags rhs) {
```
What about to make a template/macro that will automatically define operators?

jrohel replied:
In fact I wanted to discuss about it.

j-mracek commented on libdnf/rpm/repo.cpp:
```diff
@@ -223,7 +223,7 @@ const std::string & Repo::Impl::get_metadata_path(const std::string & metadata_t
     auto & ret = (it != metadata_paths.end()) ? it->second : empty;
     if (ret.empty()) {
-        logger.debug(fmt::format("not found \"{}\" for: {}", metadata_type, conf->name().get_value()));
+        logger.debug(fmt::format("not found \"{}\" for: {}", metadata_type, id));
```
Why did you change to report repoid instead of name? I think that repoid is a better better choice because it is unique but still I have an interest to know your reasons.

jrohel replied:
Exactly. The repository id is unique "name" of the repository. And it is used in libsolv too.
But what is `conf->name()`? A nicer more descriptive name. I think it is good for output to console. 
For logging I prefer unique repository id.

j-mracek commented on libdnf/rpm/repo_impl.hpp:
```diff
@@ -92,8 +80,62 @@ class Key {
+    enum class DataType { FILENAMES, PRESTO, UPDATEINFO, OTHER };
+    enum class DataState { NEW, LOADED_FETCH, LOADED_CACHE, WRITTEN };
+
+    Id get_data_id(DataType which) const noexcept;
```
Please could you add a comment what is returning. Id or int is not informative enough.

j-mracek commented on libdnf/rpm/repo_impl.hpp:
```diff
@@ -92,8 +80,62 @@ class Key {
+    // Checksum of data in .solv file. Used for validity check of .solvx files.
+    unsigned char checksum[CHKSUM_BYTES];
+
+    DataState state_main{DataState::NEW};
```
This construct is able to track only states of certain type of data. What about to store the information in map? Absent key could mean not requested.

j-mracek commented on libdnf/rpm/repo_impl.hpp:
```diff
@@ -92,8 +80,62 @@ class Key {
+    DataState state_presto{DataState::NEW};
+    DataState state_updateinfo{DataState::NEW};
+    DataState state_other{DataState::NEW};
+    Id filenames_repodata{0};
```
Please could you add a comment for what it is needed or what it represents?

j-mracek commented on libdnf/rpm/solv_sack.cpp:
```diff
@@ -0,0 +1,456 @@
+
+std::string SolvSack::Impl::give_repo_solv_cache_fn(const std::string & repoid, const char * ext) {
+    auto & cachedir = base->get_config().cachedir().get_value();
+    std::string fn = cachedir + "/" + repoid;
```
What about when cachedir == '/tmp/'

j-mracek replied:
it will create a path /tmp//<repoid>

j-mracek replied:
I think that we planned to store solv files in repo cache directory.

jrohel replied:
I don't understand the question? You are meaning doubling of `/` character. That's not problem. But fact is that I plan use C++17  `std::filesystem::path` in future. So I can do it now.

jrohel committed:
[dnf5] Remove repodata states,ids from LibsolvRepoExt, SolvSack::load_repo()

- There isn't need to store repodata states and repodata ids.
- SolvSack::load_repo() implementation code was moved to Impl.

jrohel committed:
[dnf5] solv_cache: use filesystem::path for concatenate path components

j-mracek commented:
LGTM
